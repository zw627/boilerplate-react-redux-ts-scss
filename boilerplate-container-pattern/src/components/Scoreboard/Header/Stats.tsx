import React from "react";

import { StatsProps } from "Containers/Scoreboard/Header/Stats";

const Stats: React.FC<StatsProps> = ({
  playerCount,
  scoreCount,
}: StatsProps) => {
  return (
    <table className="stats">
      <tbody>
        <tr>
          <td>Players:</td>
          <td>{playerCount}</td>
        </tr>
        <tr>
          <td>Total Points:</td>
          <td>{scoreCount}</td>
        </tr>
      </tbody>
    </table>
  );
};

export default Stats;
