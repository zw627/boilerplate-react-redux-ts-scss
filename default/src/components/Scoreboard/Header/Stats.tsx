import React from "react";
import { useSelector } from "react-redux";

import { AppState } from "Store/index";
import { scoreCountSelector, playerCountSelector } from "Utils/selectors";

const Stats: React.FC<{}> = () => {
  // State (redux)
  const scoreCount = useSelector((state: AppState) =>
    scoreCountSelector(state)
  );
  const playerCount = useSelector((state: AppState) =>
    playerCountSelector(state)
  );

  return (
    <table className="stats">
      <tbody>
        <tr>
          <td>Players:</td>
          <td>{playerCount}</td>
        </tr>
        <tr>
          <td>Total Points:</td>
          <td>{scoreCount}</td>
        </tr>
      </tbody>
    </table>
  );
};

export default Stats;
