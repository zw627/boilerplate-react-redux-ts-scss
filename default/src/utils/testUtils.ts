// Helper functions for Jest and Enzyme

import { ReactWrapper, ShallowWrapper } from "enzyme";
import { createStore, Store } from "redux";

import { rootReducer } from "Store/index";

// Type for parameters of setupStore()
type setupStorePars = {
  playerList?: {
    id: string;
    name: string;
    score: number;
    created: string;
    updated: string;
    isSelected: boolean;
  }[];
  selectedPlayerId?: string;
  isRunning?: boolean;
  previousTime?: number;
  elapsedTime?: number;
  isLightMode?: boolean;
  visibility?: boolean;
};

// Setup a store with some fake states for testing
export function setupStore({
  playerList = [
    {
      id: "1654A",
      name: "Alpha",
      score: 10,
      created: "11/21/2016",
      updated: "06/12/2018",
      isSelected: false,
    },
    {
      id: "1654B",
      name: "Beta",
      score: 4,
      created: "11/8/2016",
      updated: "11/9/2017",
      isSelected: false,
    },
  ],
  isRunning = false,
  previousTime = 0,
  elapsedTime = 0,
  isLightMode = false,
  visibility = false,
}: setupStorePars = {}): Store {
  return createStore(rootReducer, {
    notification: {
      storage: {
        visibility,
      },
    },
    scoreboard: {
      player: {
        playerList,
      },
      stopwatch: {
        isRunning,
        previousTime,
        elapsedTime,
      },
      theme: {
        isLightMode,
      },
    },
  });
}

// Multiple toEqual, useful when testing a table
export function multiToEqual([...args]: Array<string[] | number[]>): void {
  // [[a, b], [a, b], [a, b]]
  [...args].map((arg) => {
    // [a, b] (expect a equals to b)
    expect(arg[0]).toEqual(arg[1]);
  });
}

export function matchLength(
  wrapper: ReactWrapper | ShallowWrapper,
  [element, number]: [string, number]
): void {
  expect(wrapper.find(element).length).toEqual(number);
}

// Multiple matchLength
export function matchMultiLength(
  wrapper: ReactWrapper | ShallowWrapper,
  [...args]: [string, number][]
): void {
  // [[a, b], [a, b], [a, b]]
  [...args].map((arg) => {
    // [a, b] (expect a equals to b)
    matchLength(wrapper, [arg[0], arg[1]]);
  });
}

// export const matchText = (
//   wrapper: ReactWrapper | ShallowWrapper,
//   [element, index]: [string, number],
//   text: string
// ): void => {
//   expect(
//     wrapper
//       .find(element)
//       .at(index)
//       .text()
//   ).toEqual(text);
// };

// export const matchMultiText = (
//   wrapper: ReactWrapper | ShallowWrapper,
//   [...args]
// ): void => {
//   if ([...args].length > 0)
//     [...args].map(arg => {
//       matchText(wrapper, [arg[0], arg[1]], arg[2]);
//     });
// };
