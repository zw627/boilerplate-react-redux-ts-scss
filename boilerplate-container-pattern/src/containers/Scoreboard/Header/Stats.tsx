import { connect } from "react-redux";

// Component
import Stats from "Components/Scoreboard/Header/Stats";

// Store
import { playerCountSelector, scoreCountSelector } from "Utils/selectors";
import { AppState } from "Store/index";

export interface StatsProps {
  playerCount: number;
  scoreCount: number;
}

const mapStateToProps = (state: AppState): StatsProps => ({
  playerCount: playerCountSelector(state),
  scoreCount: scoreCountSelector(state),
});

export default connect(mapStateToProps)(Stats);
