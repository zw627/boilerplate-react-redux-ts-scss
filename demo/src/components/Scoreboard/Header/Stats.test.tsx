import React from "react";
import { Provider } from "react-redux";
import { Store } from "redux";
import { mount, ReactWrapper } from "enzyme";

import Stats from "Components/Scoreboard/Header/Stats";
import { setupStore, multiToEqual } from "Utils/testUtils";

function setupWrapper(mockStore = setupStore(), props = {}): ReactWrapper {
  return mount(
    <Provider store={mockStore}>
      <Stats {...props} />
    </Provider>
  ) as ReactWrapper;
}

describe("Scoreboard/Stats", () => {
  let mockStore: Store;
  let wrapper: ReactWrapper;

  // Setup store and wrapper
  beforeEach(() => {
    mockStore = setupStore();
    wrapper = setupWrapper(mockStore, {});
  });

  test("compare to the last snapshot", () => {
    expect(wrapper.find(Stats)).toMatchSnapshot();
  });

  it("should display player and point count", () => {
    function mapRowColumns(index: number): string[] {
      return wrapper
        .find("tr")
        .at(index)
        .find("td")
        .map((column) => column.text());
    }
    const firstRowColumns = mapRowColumns(0);
    const secondRowColumns = mapRowColumns(1);
    /* eslint jest/expect-expect: ["error", { "assertFunctionNames": ["expect", "multiToEqual"] }] */
    multiToEqual([
      [firstRowColumns.length, 2],
      [firstRowColumns[0], "Players:"],
      [firstRowColumns[1], "2"],
      [secondRowColumns.length, 2],
      [secondRowColumns[0], "Total Points:"],
      [secondRowColumns[1], "14"],
    ]);
  });
});
