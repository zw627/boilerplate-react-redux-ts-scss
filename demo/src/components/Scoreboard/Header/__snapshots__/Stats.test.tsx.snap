// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scoreboard/Stats compare to the last snapshot 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Stats>
    <table
      className="stats"
    >
      <tbody>
        <tr>
          <td>
            Players:
          </td>
          <td>
            2
          </td>
        </tr>
        <tr>
          <td>
            Total Points:
          </td>
          <td>
            14
          </td>
        </tr>
      </tbody>
    </table>
  </Stats>
</Provider>
`;
